stdlib/mlang/composition-check.mc

```diff
@@ -60,15 +69,9 @@ let collectPats = lam env. lam includes.
   in 
   join (map incl2pats includes)
 
-let carthesianProduct : all a. all b. [a] -> [b] -> [(a, b)] = lam xs. lam ys.
-  let work1 : a -> [b] -> [(a, b)] = lam x. lam ys. 
-    map (lam y. (x, y)) ys
-  in 
-  join (map (lam x. work1 x ys) xs)
-
 let indexPairs : Int -> [(Int, Int)] = lam n. 
   let indices = range 0 n 1 in 
-  let pairs = carthesianProduct indices indices in
+  let pairs = seqLiftA2 (lam a. lam b. (a, b)) indices indices in
```

Comment:
This does at least twice the work it needs because of the filter below. I would suggest you port https://github.com/miking-lang/miking/blob/dcdbdab33a52312d9e0195dcad994b81c1e59854/src/boot/lib/mlang.ml#L26-L30 and put it in `seq.mc` (your choice if it should create the pair or take a function to call for each pair).

Reply:
done.